File: module/PromptBuilder.py
import threading
from functools import lru_cache

from base.Base import Base
from base.BaseLanguage import BaseLanguage
from model.Item import Item
from module.Config import Config
from module.Data.DataManager import DataManager
from module.QualityRule.QualityRuleSnapshot import QualityRuleSnapshot
from module.Utils.JSONTool import JSONTool


class PromptBuilder(Base):
    # 类线程锁
    LOCK: threading.Lock = threading.Lock()

    # "原文/Source" 提示词占位文本（用于 source_language=ALL 时）
    SOURCE_PLACEHOLDER_ZH: str = "原文"
    SOURCE_PLACEHOLDER_EN: str = "Source"

    def __init__(
        self, config: Config, quality_snapshot: QualityRuleSnapshot | None = None
    ) -> None:
        super().__init__()

        # 初始化
        self.config: Config = config
        self.quality_snapshot: QualityRuleSnapshot | None = quality_snapshot

    @classmethod
    def reset(cls) -> None:
        cls.get_base.cache_clear()
        cls.get_prefix.cache_clear()
        cls.get_suffix.cache_clear()
        cls.get_suffix_thinking.cache_clear()
        cls.get_suffix_glossary.cache_clear()
        cls.get_analysis_base.cache_clear()
        cls.get_analysis_prefix.cache_clear()
        cls.get_analysis_suffix.cache_clear()

    @classmethod
    def read_prompt_text(
        cls, prompt_group: str, language: BaseLanguage.Enum, file_name: str
    ) -> str:
        with open(
            f"resource/preset/{prompt_group}/{language.lower()}/{file_name}",
            "r",
            encoding="utf-8-sig",
        ) as reader:
            return reader.read().strip()

    @classmethod
    @lru_cache(maxsize=None)
    def get_base(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt", language, "base.txt")

    @classmethod
    @lru_cache(maxsize=None)
    def get_prefix(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt", language, "prefix.txt")

    @classmethod
    @lru_cache(maxsize=None)
    def get_suffix(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt", language, "suffix.txt")

    @classmethod
    @lru_cache(maxsize=None)
    def get_suffix_thinking(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt", language, "thinking.txt")

    @classmethod
    @lru_cache(maxsize=None)
    def get_suffix_glossary(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt", language, "suffix_glossary.txt")

    @classmethod
    @lru_cache(maxsize=None)
    def get_analysis_base(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt_glossary", language, "base.txt")

    @classmethod
    @lru_cache(maxsize=None)
    def get_analysis_prefix(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt_glossary", language, "prefix.txt")

    @classmethod
    @lru_cache(maxsize=None)
    def get_analysis_suffix(cls, language: BaseLanguage.Enum) -> str:
        return cls.read_prompt_text("prompt_glossary", language, "suffix.txt")

    def resolve_prompt_languages(
        self,
    ) -> tuple[BaseLanguage.Enum, str, str, str]:
        """统一推导提示词语言与占位文本，避免翻译/分析两条链重复分叉。"""
        languages = BaseLanguage.get_languages()

        if self.config.target_language == BaseLanguage.ALL:
            raise ValueError("target_language does not support ALL")
        if self.config.target_language not in languages:
            raise ValueError(f"invalid target_language: {self.config.target_language}")

        if self.config.target_language == BaseLanguage.Enum.ZH:
            prompt_language = BaseLanguage.Enum.ZH
            source_placeholder = __class__.SOURCE_PLACEHOLDER_ZH
            if self.config.source_language in languages:
                source_language = BaseLanguage.get_name_zh(self.config.source_language)
            else:
                source_language = source_placeholder
            target_language = BaseLanguage.get_name_zh(self.config.target_language)
        else:
            prompt_language = BaseLanguage.Enum.EN
            source_placeholder = __class__.SOURCE_PLACEHOLDER_EN
            if self.config.source_language in languages:
                source_language = BaseLanguage.get_name_en(self.config.source_language)
            else:
                source_language = source_placeholder
            target_language = BaseLanguage.get_name_en(self.config.target_language)

        if not source_language:
            source_language = source_placeholder
        if not target_language:
            raise ValueError(f"invalid target_language: {self.config.target_language}")

        return prompt_language, source_placeholder, source_language, target_language

    # 获取自定义提示词数据
    def get_custom_prompt_data(self, language: BaseLanguage.Enum) -> str:
        snapshot = self.quality_snapshot
        if snapshot is not None:
            if language == BaseLanguage.Enum.ZH:
                return snapshot.custom_prompt_zh
            return snapshot.custom_prompt_en

        if language == BaseLanguage.Enum.ZH:
            return DataManager.get().get_custom_prompt_zh()
        return DataManager.get().get_custom_prompt_en()

    # 获取自定义提示词启用状态
    def get_custom_prompt_enable(self, language: BaseLanguage.Enum) -> bool:
        snapshot = self.quality_snapshot
        if snapshot is not None:
            if language == BaseLanguage.Enum.ZH:
                return snapshot.custom_prompt_zh_enable
            return snapshot.custom_prompt_en_enable

        if language == BaseLanguage.Enum.ZH:
            return DataManager.get().get_custom_prompt_zh_enable()
        return DataManager.get().get_custom_prompt_en_enable()

    def resolve_main_prompt_base(self, prompt_language: BaseLanguage.Enum) -> str:
        """统一决定主提示词正文来源，避免中英文分支重复判断。"""
        if self.get_custom_prompt_enable(prompt_language):
            return self.get_custom_prompt_data(prompt_language)
        return __class__.get_base(prompt_language)

    # 获取主提示词
    def build_main(self) -> str:
        prompt_language, _source_placeholder, source_language, target_language = (
            self.resolve_prompt_languages()
        )

        with __class__.LOCK:
            # 前缀
            prefix = __class__.get_prefix(prompt_language)

            # 主体
            base = self.resolve_main_prompt_base(prompt_language)

            # 思考块：与输出块分离，避免自动术语表切换时互相覆盖
            thinking = ""
            if self.config.force_thinking_enable:
                thinking = __class__.get_suffix_thinking(prompt_language)

            # 输出块
            if not self.config.auto_glossary_enable:
                suffix_output = __class__.get_suffix(prompt_language)
            else:
                suffix_output = __class__.get_suffix_glossary(prompt_language)

        # 组装提示词：输出块必须位于末尾，避免影响 JSONLINE 规则
        base_block = "\n".join([prefix, base])
        parts = [base_block]
        if thinking:
            parts.append(thinking)
        parts.append(suffix_output)
        full_prompt = "\n\n".join(parts)
        full_prompt = full_prompt.replace("{source_language}", source_language)
        full_prompt = full_prompt.replace("{target_language}", target_language)

        return full_prompt

    def build_glossary_analysis_main(self) -> str:
        """构建术语分析任务的主提示词。"""
        prompt_language, _source_placeholder, _source_language, target_language = (
            self.resolve_prompt_languages()
        )

        with __class__.LOCK:
            prefix = __class__.get_analysis_prefix(prompt_language)
            base = __class__.get_analysis_base(prompt_language)
            suffix = __class__.get_analysis_suffix(prompt_language)

        full_prompt = "\n\n".join([prefix, base, suffix])
        return full_prompt.replace("{target_language}", target_language)

    # 构造参考上文
    def build_preceding(self, precedings: list[Item]) -> str:
        if not precedings:
            return ""
        elif self.config.target_language == BaseLanguage.Enum.ZH:
            return (
                "参考上文："
                + "\n"
                + "\n".join(
                    [item.get_src().strip().replace("\n", "\\n") for item in precedings]
                )
            )
        else:
            return (
                "Preceding Context:"
                + "\n"
                + "\n".join(
                    [item.get_src().strip().replace("\n", "\\n") for item in precedings]
                )
            )

    # 构造术语表
    def build_glossary(self, srcs: list[str]) -> str:
        full = "\n".join(srcs)
        full_lower = full.lower()  # 用于不区分大小写的匹配

        # 筛选匹配的术语
        glossary: list[dict[str, str]] = []
        glossary_data = (
            self.quality_snapshot.get_glossary_entries()
            if self.quality_snapshot is not None
            else DataManager.get().get_glossary()
        )

        for v in glossary_data:
            src = v.get("src", "")
            is_case_sensitive = v.get("case_sensitive", False)

            # 根据 case_sensitive 决定匹配方式
            if is_case_sensitive:
                # 大小写敏感：直接使用 in
                if src in full:
                    glossary.append(v)
            else:
                # 大小写不敏感：转换为小写后匹配
                if src.lower() in full_lower:
                    glossary.append(v)

        # 构建文本
        result = []
        for item in glossary:
            src = item.get("src", "")
            dst = item.get("dst", "")
            info = item.get("info", "")

            if info == "":
                result.append(f"{src} -> {dst}")
            else:
                result.append(f"{src} -> {dst} #{info}")

        # 返回结果
        if not result:
            return ""

        if self.config.target_language == BaseLanguage.Enum.ZH:
            return (
                "术语表 <术语原文> -> <术语译文> #<术语信息>:"
                + "\n"
                + "\n".join(result)
            )
        else:
            return (
                "Glossary <Original Term> -> <Translated Term> #<Term Information>:"
                + "\n"
                + "\n".join(result)
            )

    # 构造术语表
    def build_glossary_sakura(self, srcs: list[str]) -> str:
        full = "\n".join(srcs)
        full_lower = full.lower()  # 用于不区分大小写的匹配

        # 筛选匹配的术语
        glossary: list[dict[str, str]] = []
        glossary_data = (
            self.quality_snapshot.get_glossary_entries()
            if self.quality_snapshot is not None
            else DataManager.get().get_glossary()
        )

        for v in glossary_data:
            src = v.get("src", "")
            is_case_sensitive = v.get("case_sensitive", False)

            # 根据 case_sensitive 决定匹配方式
            if is_case_sensitive:
                # 大小写敏感：直接使用 in
                if src in full:
                    glossary.append(v)
            else:
                # 大小写不敏感：转换为小写后匹配
                if src.lower() in full_lower:
                    glossary.append(v)

        # 构建文本
        result = []
        for item in glossary:
            src = item.get("src", "")
            dst = item.get("dst", "")
            info = item.get("info", "")

            if info == "":
                result.append(f"{src}->{dst}")
            else:
                result.append(f"{src}->{dst} #{info}")

        # 返回结果
        if not result:
            return ""

        return "\n".join(result)

    # 构建控制字符示例
    def build_control_characters_samples(self, main: str, samples: list[str]) -> str:
        # 去重并过滤空字符串
        unique_samples = {v.strip() for v in samples if v.strip()}

        if not unique_samples:
            return ""

        # 只有在提示词中明确包含 控制符 时才添加相应示例
        main_lower = main.lower()
        has_instruction = (
            "控制符" in main
            or "控制字符" in main
            or "control code" in main_lower
            or "control character" in main_lower
        )
        if not has_instruction:
            return ""

        # 判断提示词语言
        if self.config.target_language == BaseLanguage.Enum.ZH:
            prefix: str = "控制字符示例："
        else:
            prefix: str = "Control Characters Samples:"

        return prefix + "\n" + f"{', '.join(unique_samples)}"

    # 构建输入
    def build_inputs(self, srcs: list[str]) -> str:
        inputs = "\n".join(
            JSONTool.dumps({str(i): line}) for i, line in enumerate(srcs)
        )

        if self.config.target_language == BaseLanguage.Enum.ZH:
            return "输入：\n" + "```jsonline\n" + f"{inputs}\n" + "```"
        else:
            return "Input:\n" + "```jsonline\n" + f"{inputs}\n" + "```"

    def build_analysis_inputs(self, srcs: list[str]) -> str:
        """分析任务只需要纯文本原文，避免额外结构干扰模型抽取术语。"""
        if not srcs:
            return ""

        inputs = "\n".join(srcs)
        if self.config.target_language == BaseLanguage.Enum.ZH:
            return "输入：\n" + inputs
        else:
            return "Input:\n" + inputs

    # 生成提示词
    def generate_prompt(
        self,
        srcs: list[str],
        samples: list[str],
        precedings: list[Item],
    ) -> tuple[list[dict], list[str]]:
        # 初始化
        messages: list[dict[str, str]] = []
        console_log: list[str] = []

        # system=稳定指令（规则/格式约束）；user=本次任务动态数据。
        instruction_text = self.build_main()
        user_parts: list[str] = []

        # 参考上文
        result = self.build_preceding(precedings)
        if result != "":
            user_parts.append(result)
            console_log.append(result)

        # 术语表
        glossary_enable = (
            self.quality_snapshot.glossary_enable
            if self.quality_snapshot is not None
            else DataManager.get().get_glossary_enable()
        )
        if glossary_enable:
            result = self.build_glossary(srcs)
            if result != "":
                user_parts.append(result)

                console_log.append(result)

        # 控制字符示例
        # 触发条件只检查 system 指令文本，避免 user 数据误触发。
        result = self.build_control_characters_samples(instruction_text, samples)
        if result != "":
            user_parts.append(result)
            console_log.append(result)

        # 输入
        result = self.build_inputs(srcs)
        if result != "":
            user_parts.append(result)

        messages.append({"role": "system", "content": instruction_text})
        messages.append({"role": "user", "content": "\n\n".join(user_parts)})

        return messages, console_log

    # 生成提示词 - Sakura
    def generate_prompt_sakura(self, srcs: list[str]) -> tuple[list[dict], list[str]]:
        # 初始化
        messages: list[dict[str, str]] = []
        console_log: list[str] = []

        # 构建系统提示词
        messages.append(
            {
                "role": "system",
                "content": "你是一个轻小说翻译模型，可以流畅通顺地以日本轻小说的风格将日文翻译成简体中文，并联系上下文正确使用人称代词，不擅自添加原文中没有的代词。",
            }
        )

        # 术语表
        content = "将下面的日文文本翻译成中文：\n" + "\n".join(srcs)
        glossary_enable = (
            self.quality_snapshot.glossary_enable
            if self.quality_snapshot is not None
            else DataManager.get().get_glossary_enable()
        )
        if glossary_enable:
            result = self.build_glossary_sakura(srcs)
            if result != "":
                content = (
                    "根据以下术语表（可以为空）：\n"
                    + result
                    + "\n"
                    + "将下面的日文文本根据对应关系和备注翻译成中文：\n"
                    + "\n".join(srcs)
                )
                console_log.append(result)

        # 构建提示词列表
        messages.append(
            {
                "role": "user",
                "content": content,
            }
        )

        return messages, console_log

    def generate_glossary_prompt(
        self,
        srcs: list[str],
    ) -> tuple[list[dict[str, str]], list[str]]:
        """生成术语分析任务提示词。

        为什么单独建方法：
        - 分析任务使用独立提示词模板，不应混入翻译专用的控制符说明
        - 分析任务只看当前原文，避免上文和已有术语影响抽取结果
        """

        messages: list[dict[str, str]] = []
        console_log: list[str] = []

        instruction_text = self.build_glossary_analysis_main()
        inputs_text = self.build_analysis_inputs(srcs)

        messages.append({"role": "system", "content": instruction_text})
        messages.append({"role": "user", "content": inputs_text})
        return messages, console_log

File: src/main/task-worker/prompt/prompt-builder.ts
import path from "node:path";
import { readFile } from "node:fs/promises";

import { JsonTool } from "../../../shared/utils/json-tool";
import type { TextQualitySnapshot, TextTaskItemRecord } from "../../../shared/text/text-types";
import type { LlmRequestMessage } from "../llm/llm-types";

// 中文提示词模板使用“原文”占位，英文模板使用“Source”占位，避免字符串散落在构造逻辑里。
const SOURCE_PLACEHOLDER_ZH = "原文";
// 英文模板占位符单独保留，避免后续模板本地化时误改中文占位。
const SOURCE_PLACEHOLDER_EN = "Source";

// 中文 UI 下展示的语言名，直接写入模型提示词。
const LANGUAGE_NAME_ZH: Record<string, string> = {
  ZH: "简体中文",
  EN: "英语",
  JA: "日语",
  KO: "韩语",
  RU: "俄语",
  AR: "阿拉伯语",
  DE: "德语",
  FR: "法语",
  PL: "波兰语",
  ES: "西班牙语",
  IT: "意大利语",
  PT: "葡萄牙语",
  HU: "匈牙利语",
  TR: "土耳其语",
  TH: "泰语",
  ID: "印度尼西亚语",
  VI: "越南语",
};

// 非中文 UI 下展示的语言名，保持和资源模板英文语境一致。
const LANGUAGE_NAME_EN: Record<string, string> = {
  ZH: "Simplified Chinese",
  EN: "English",
  JA: "Japanese",
  KO: "Korean",
  RU: "Russian",
  AR: "Arabic",
  DE: "German",
  FR: "French",
  PL: "Polish",
  ES: "Spanish",
  IT: "Italian",
  PT: "Portuguese",
  HU: "Hungarian",
  TR: "Turkish",
  TH: "Thai",
  ID: "Indonesian",
  VI: "Vietnamese",
};

/**
 * 提示词构造所需的最小配置快照，worker 只读取语言与界面语言。
 */
export interface PromptBuilderConfig {
  app_language?: string;
  source_language?: string;
  target_language?: string;
}

/**
 * PromptBuilder 输出给 LLM adapter 的消息和本地日志展示文本。
 */
export interface PromptBuildResult {
  messages: LlmRequestMessage[];
  console_log: string[];
}

/**
 * worker 侧提示词构造器，读取资源模板并拼接本次 work unit 动态数据。
 */
export class PromptBuilder {
  private readonly app_root: string;
  private readonly config: PromptBuilderConfig;
  private readonly quality_snapshot: TextQualitySnapshot;

  /**
   * app_root 由 Electron main 注入，worker 不自行猜测资源根。
   */
  public constructor(
    app_root: string,
    config: PromptBuilderConfig,
    quality_snapshot: TextQualitySnapshot,
  ) {
    this.app_root = app_root;
    this.config = config;
    this.quality_snapshot = quality_snapshot;
  }

  /**
   * 生成普通翻译提示词；system 放稳定指令，user 放本次输入和术语。
   */
  public async generate_prompt(
    srcs: string[],
    samples: string[],
    precedings: TextTaskItemRecord[],
  ): Promise<PromptBuildResult> {
    const messages: LlmRequestMessage[] = [];
    const console_log: string[] = [];
    const instruction_text = await this.build_main();
    const user_parts: string[] = [];

    const preceding = this.build_preceding(precedings);
    if (preceding !== "") {
      user_parts.push(preceding);
      console_log.push(preceding);
    }

    if (this.quality_snapshot.glossary_enable) {
      const glossary = this.build_glossary(srcs);
      if (glossary !== "") {
        user_parts.push(glossary);
        console_log.push(glossary);
      }
    }

    const control_samples = this.build_control_characters_samples(instruction_text, samples);
    if (control_samples !== "") {
      user_parts.push(control_samples);
      console_log.push(control_samples);
    }

    const inputs = this.build_inputs(srcs);
    if (inputs !== "") {
      user_parts.push(inputs);
    }

    messages.push({ role: "system", content: instruction_text });
    messages.push({ role: "user", content: user_parts.join("\n\n") });
    return { messages, console_log };
  }

  /**
   * 生成 SakuraLLM 固定提示词，保持旧模型专用语义。
   */
  public generate_prompt_sakura(srcs: string[]): PromptBuildResult {
    const messages: LlmRequestMessage[] = [
      {
        role: "system",
        content:
          "你是一个轻小说翻译模型，可以流畅通顺地以日本轻小说的风格将日文翻译成简体中文，并联系上下文正确使用人称代词，不擅自添加原文中没有的代词。",
      },
    ];
    const console_log: string[] = [];
    let content = `将下面的日文文本翻译成中文：\n${srcs.join("\n")}`;
    if (this.quality_snapshot.glossary_enable) {
      const glossary = this.build_glossary_sakura(srcs);
      if (glossary !== "") {
        content = `根据以下术语表（可以为空）：\n${glossary}\n将下面的日文文本根据对应关系和备注翻译成中文：\n${srcs.join("\n")}`;
        console_log.push(glossary);
      }
    }
    messages.push({ role: "user", content });
    return { messages, console_log };
  }

  /**
   * 生成术语分析提示词；分析链路不混入上文或翻译控制字符示例。
   */
  public async generate_glossary_prompt(srcs: string[]): Promise<PromptBuildResult> {
    const instruction_text = await this.build_glossary_analysis_main();
    const inputs_text = this.build_analysis_inputs(srcs);
    return {
      messages: [
        { role: "system", content: instruction_text },
        { role: "user", content: inputs_text },
      ],
      console_log: [],
    };
  }

  /**
   * 翻译主提示词从自定义快照或资源模板读取。
   */
  public async build_main(): Promise<string> {
    const context = this.resolve_prompt_context();
    const prefix = await this.read_prompt_text(
      "translation_prompt",
      context.prompt_language,
      "prefix.txt",
    );
    const base = this.quality_snapshot.translation_prompt_enable
      ? this.quality_snapshot.translation_prompt
      : await this.read_prompt_text("translation_prompt", context.prompt_language, "base.txt");
    const thinking = await this.read_prompt_text(
      "translation_prompt",
      context.prompt_language,
      "thinking.txt",
    );
    const suffix = await this.read_prompt_text(
      "translation_prompt",
      context.prompt_language,
      "suffix.txt",
    );
    return this.join_prompt_sections(prefix, base, thinking, suffix)
      .replaceAll("{source_language}", context.source_language)
      .replaceAll("{target_language}", context.target_language);
  }

  /**
   * 分析主提示词只替换目标语言，不携带源语言占位。
   */
  public async build_glossary_analysis_main(): Promise<string> {
    const context = this.resolve_prompt_context();
    const prefix = await this.read_prompt_text(
      "analysis_prompt",
      context.prompt_language,
      "prefix.txt",
    );
    const base = this.quality_snapshot.analysis_prompt_enable
      ? this.quality_snapshot.analysis_prompt
      : await this.read_prompt_text("analysis_prompt", context.prompt_language, "base.txt");
    const thinking = await this.read_prompt_text(
      "analysis_prompt",
      context.prompt_language,
      "thinking.txt",
    );
    const suffix = await this.read_prompt_text(
      "analysis_prompt",
      context.prompt_language,
      "suffix.txt",
    );
    return this.join_prompt_sections(prefix, base, thinking, suffix).replaceAll(
      "{target_language}",
      context.target_language,
    );
  }

  /**
   * 参考上文只放 user prompt，避免系统指令随上下文变化。
   */
  public build_preceding(precedings: TextTaskItemRecord[]): string {
    if (precedings.length === 0) {
      return "";
    }
    const lines = precedings.map((item) =>
      String(item.src ?? "")
        .trim()
        .replaceAll("\n", "\\n"),
    );
    return this.is_prompt_ui_zh()
      ? `参考上文：\n${lines.join("\n")}`
      : `Preceding Context:\n${lines.join("\n")}`;
  }

  /**
   * 术语表按当前输入全文命中过滤，未命中时不污染 prompt。
   */
  public build_glossary(srcs: string[]): string {
    const result = this.build_glossary_lines(srcs, " -> ");
    if (result.length === 0) {
      return "";
    }
    return this.is_prompt_ui_zh()
      ? `术语表 <术语原文> -> <术语译文> #<术语信息>:\n${result.join("\n")}`
      : `Glossary <Original Term> -> <Translated Term> #<Term Information>:\n${result.join("\n")}`;
  }

  /**
   * SakuraLLM 术语格式不带空格，保持旧提示词格式。
   */
  public build_glossary_sakura(srcs: string[]): string {
    return this.build_glossary_lines(srcs, "->").join("\n");
  }

  /**
   * 控制字符示例只在系统提示词明确要求控制符时加入。
   */
  public build_control_characters_samples(main: string, samples: string[]): string {
    const unique_samples = [...new Set(samples.map((sample) => sample.trim()).filter(Boolean))];
    if (unique_samples.length === 0) {
      return "";
    }
    const main_lower = main.toLowerCase();
    if (
      !(
        main.includes("控制符") ||
        main.includes("控制字符") ||
        main_lower.includes("control code") ||
        main_lower.includes("control character")
      )
    ) {
      return "";
    }
    return this.is_prompt_ui_zh()
      ? `控制字符示例：\n${unique_samples.join(", ")}`
      : `Control Characters Samples:\n${unique_samples.join(", ")}`;
  }

  /**
   * 翻译输入固定为 jsonline，响应解码器也按此格式优先解析。
   */
  public build_inputs(srcs: string[]): string {
    const inputs = srcs
      .map((line, index) => JsonTool.stringifyStrict({ [String(index)]: line }))
      .join("\n");
    return this.is_prompt_ui_zh()
      ? `输入：\n\`\`\`jsonline\n${inputs}\n\`\`\``
      : `Input:\n\`\`\`jsonline\n${inputs}\n\`\`\``;
  }

  /**
   * 分析输入保持纯文本，减少模型把 JSON key 当作术语。
   */
  public build_analysis_inputs(srcs: string[]): string {
    if (srcs.length === 0) {
      return "";
    }
    return this.is_prompt_ui_zh() ? `输入：\n${srcs.join("\n")}` : `Input:\n${srcs.join("\n")}`;
  }

  /**
   * 模板段落拼接统一在这里处理，保证输出约束始终位于最后。
   */
  public join_prompt_sections(
    prefix: string,
    base: string,
    thinking: string,
    suffix: string,
  ): string {
    const parts = [`${prefix}\n${base}`];
    if (thinking !== "") {
      parts.push(thinking);
    }
    parts.push(suffix);
    return parts.join("\n\n");
  }

  /**
   * UI 语言只支持中英提示词模板，未知值回退中文。
   */
  private get_prompt_ui_language(): "zh" | "en" {
    return String(this.config.app_language ?? "ZH").toUpperCase() === "EN" ? "en" : "zh";
  }

  /**
   * 中文 UI 决定提示词标题和语言名显示口径。
   */
  private is_prompt_ui_zh(): boolean {
    return this.get_prompt_ui_language() === "zh";
  }

  /**
   * 解析提示词语言、源语言占位和目标语言名。
   */
  private resolve_prompt_context(): {
    prompt_language: "zh" | "en";
    source_language: string;
    target_language: string;
  } {
    const prompt_language = this.get_prompt_ui_language();
    const source_code = String(this.config.source_language ?? "ALL").toUpperCase();
    const target_code = String(this.config.target_language ?? "ZH").toUpperCase();
    const names = prompt_language === "zh" ? LANGUAGE_NAME_ZH : LANGUAGE_NAME_EN;
    const source_placeholder =
      prompt_language === "zh" ? SOURCE_PLACEHOLDER_ZH : SOURCE_PLACEHOLDER_EN;
    return {
      prompt_language,
      source_language: names[source_code] ?? source_placeholder,
      target_language: names[target_code] ?? names.ZH,
    };
  }

  /**
   * 模板路径固定在 resource 下，worker 不读用户预设目录。
   */
  private async read_prompt_text(
    task_dir_name: string,
    language: "zh" | "en",
    file_name: string,
  ): Promise<string> {
    const template_path = path.join(
      this.app_root,
      "resource",
      task_dir_name,
      "template",
      language,
      file_name,
    );
    return (await readFile(template_path, "utf-8")).trim();
  }

  /**
   * 术语匹配尊重大小写标志，命中后按指定分隔符生成行文本。
   */
  private build_glossary_lines(srcs: string[], separator: string): string[] {
    const full_text = srcs.join("\n");
    const full_text_lower = full_text.toLowerCase();
    const result: string[] = [];
    for (const entry of this.quality_snapshot.glossary_entries) {
      const src = String(entry["src"] ?? "");
      const dst = String(entry["dst"] ?? "");
      const info = String(entry["info"] ?? "");
      const case_sensitive = entry["case_sensitive"] === true;
      const matched = case_sensitive
        ? full_text.includes(src)
        : full_text_lower.includes(src.toLowerCase());
      if (!matched) {
        continue;
      }
      result.push(info === "" ? `${src}${separator}${dst}` : `${src}${separator}${dst} #${info}`);
    }
    return result;
  }
}
